src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,203 @@
+
+            var canonicalRequest = string.Join("\n", "POST", "/", "", canonicalHeaders, "", signedHeaders, Hash(body));
+
+            // Create the string to sign.
+
+            var algorithm = "AWS4-HMAC-SHA256";
+            var credentialScope = $"{datestamp}/{region}/{service}/aws4_request";
+            var stringToSign = string.Join("\n", algorithm, timestamp, credentialScope, Hash(canonicalRequest));
+
+            // Calculate the signature.
+
+            var signature = GetSignature(stringToSign, secretKey, datestamp, region, service);
+
+            // Add signing information to the request.
+
+            var authHeader = $"{algorithm} Credential={accessKey}/{credentialScope}, SignedHeaders={signedHeaders}, Signature={signature}";
+
+            return new Tuple<string, string>(authHeader, timestamp);
+        }
+
+        private static SortedDictionary<string, string> GetRequestHeaders(
+            string body,
+            string contentType,
+            string host,
+            string timestamp,
+            string sessionToken,
+            byte[] nonce)
+        {
+            var requestHeaders = new SortedDictionary<string, string>();
+            requestHeaders["Content-Type"] = contentType;
+            requestHeaders["Content-Length"] = body.Length.ToString();
+            requestHeaders["Host"] = host;
+            requestHeaders["X-Amz-Date"] = timestamp;
+            requestHeaders["X-MongoDB-Server-Nonce"] = Convert.ToBase64String(nonce);
+            requestHeaders["X-MongoDB-GS2-CB-Flag"] = "n";
+            if (sessionToken != null)
+            {
+                requestHeaders["X-Amz-Security-Token"] = sessionToken;
+            }
+
+            return requestHeaders;
+        }
+
+        private static string GetSignature(string stringToSign, string secret, string date, string region, string service)
+        {
+            byte[] kDateBlock = Hmac256(Encoding.ASCII.GetBytes("AWS4" + secret), Encoding.ASCII.GetBytes(date));
+            byte[] kRegionBlock = Hmac256(kDateBlock, Encoding.ASCII.GetBytes(region));
+            byte[] kServiceBlock = Hmac256(kRegionBlock, Encoding.ASCII.GetBytes(service));
+            byte[] kSigningBlock = Hmac256(kServiceBlock, Encoding.ASCII.GetBytes("aws4_request"));
+
+            return ToHexString(Hmac256(kSigningBlock, Encoding.ASCII.GetBytes(stringToSign)));
+        }
+
+        private static string Hash(string str)
+        {
+            byte[] bytes = Encoding.ASCII.GetBytes(str);
+            using (SHA256 algorithm = SHA256.Create())
+            {
+                var hash = algorithm.ComputeHash(bytes);
+
+                return ToHexString(hash);
+            }
+        }
+
+        private static byte[] Hmac256(byte[] keyBytes, byte[] bytes)
+        {
+            using (var hmac = new HMACSHA256(keyBytes))
+            {
+                return hmac.ComputeHash(bytes);
+            }
+        }
+
+        private static string ToHexString(byte[] bytes)
```

Comment:
NOTE: Idk what is the difference, but we have our own version of this logic: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Bson/BsonUtils.cs#L118

Reply:
Actually, no difference, I would rather use BsonUtils to avoid code duplication.
Done.